lib/SILGen/SILGenExpr.cpp

```diff
@@ -2351,9 +2350,11 @@ static bool mayLieAboutNonOptionalReturn(Expr *expr) {
     return mayLieAboutNonOptionalReturn(declRef->getDecl());
   }
 
-  // An application, which we look through to get the function we're calling.
+  // An application that produces a reference type, which we look through to
+  // get the function we're calling.
   if (auto apply = dyn_cast<ApplyExpr>(expr)) {
-    return mayLieAboutNonOptionalReturn(apply->getFn());
+    return apply->getType()->hasReferenceSemantics() &&
```

Comment:
Actually, `mayLieAboutNonOptionalReturn` already does this check; it just does it incorrectly because it doesn't notice that the decl might be a member. (And it also gets it wrong in the same way.)